blog/_posts/2016-10-06-julia-0.5-highlights.md

```diff
@@ -0,0 +1,804 @@
+As of 0.5 it still works.)
+We can verify that each tuple adds up to 100 using a comprehension to make an array of the sums of each tuple and passing that array to the `extrema` function, to find the minimum and maximum over all sums:
+
+{% highlight julia %}
+julia> extrema([sum(t) for t in change])
+(100,100)
+{% endhighlight %}
+
+Since the minimum and maximum are both 100, the sums must all be exactly 100.
+So, we know there are 242 ways to make a dollar with common coins.
+But suppose we want to ensure that the value in pennies is less than the value in nickels, and so forth.
+By adding a filter clause, we can do this easily too:
+
+{% highlight julia %}
+julia> [(q,d,n,p) for q=0:25:100 for d=0:10:100-q for n=0:5:100-q-d for p=100-q-d-n if p < n < d < q]
+4-element Array{NTuple{4,Int64},1}:
+ (50,30,15,5)
+ (50,30,20,0)
+ (50,40,10,0)
+ (75,20,5,0)
+{% endhighlight %}
+
+The only difference here is the `if p < n < d < q` clause at the end of the comprehension, which has the effect that the result only contains cases where the predicate holds true.
+There are exactly four ways make a dollar with strictly increasing value from pennies to nickels to dimes to quarters.
+
+Nested and filtered comprehensions aren't earth-shattering features – everything you can do with them can be done in a variety of other ways – but they are expressive and convenient, found in other languages, and they allow you to try more things with your data quickly and easily, with less tedious refactoring.
+
+## Generators
+
+In the previous section we used an array comprehension to take the sum of each tuple, save the sums as an array, and then pass that array of sums to the `extrema` function to find the largest and smallest sum (they should all be 100):
+
+{% highlight julia %}
+julia> extrema([sum(t) for t in change])
+(100,100)
+{% endhighlight %}
+
+But allocating an array of sums here is unnecessary: the minimum and maximum can be computed over streamed data, simply by keeping the largest and smallest values seen so far.
+In other words, this calculation could be expressed with constant memory overhead by interleaving the production of values with computation of extreme values.
+In Julia 0.4, the most straightforward way to express such an interleaved computation was to manually perform the interleaving (aka "cut and paste").
+In 0.5, if you simply omit the square brackets around an array comprehension, you get a *generator expression*, which instead of producing an array of values, can be iterated over, yielding one value at a time.
+Since `extrema` works with arbitrary iterable objects – including generators – expressing an interleaved calculation using constant memory is now as simple as deleting `[` and `]`:
+
+{% highlight julia %}
+julia> extrema(sum(t) for t in change)
+(100,100)
+{% endhighlight %}
+
```

Comment:
As I prepare the notebook, I realized that these examples are nice to show with `@time`, since the memory use in 0.5 is significantly lesser, showing that it is a generator.